Make-a-pull-request.md

```diff
@@ -0,0 +1,168 @@
+
+**Note:** If your change involves more than around 500 lines of code, we recommend first creating a short [design doc](https://github.com/oppia/oppia/wiki/Writing-design-docs). This helps avoid duplication of effort, and allows us to offer advice and suggestions on the implementation approach.
+
+To make code changes, please follow the following instructions carefully! Otherwise, your code review may be delayed.
+
+1. **Before writing any code, choose a descriptive branch name** that is lowercase and hyphen-separated, e.g. `fuzzy-rules`, and create a new branch with this name, from the current `develop` branch. Make sure that your branch name doesn't start with `develop`, `release` or `test`.
+
+   Here's how you can get the latest code from `develop`:
+
+   ```console
+   git fetch upstream
+   git checkout develop
+   git merge upstream/develop
+   ```
+
+   This will create and check out a new branch named `your-branch-name`:
+
+   ```console
+   git checkout -b your-branch-name
+   ```
+
+2. **Make commit(s) locally to your feature branch.** Each commit should be self-contained and have a descriptive commit message that helps other developers understand why the changes were made.
+
+   However, **do not write "Fix #ISSUE_NUMBER"** (e.g. "Fix #99999") in your commit messages (or any of the other closing phrases from [GitHub's documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)), as this will cause Github to close the original issue automatically.  We configure the automatic closing of issues from PRs instead of commits.
+
+   You can change your most recent commit message using `git commit --amend`. **It is difficult to change any commit messages other than your most recent one or messages on commits that have been pushed, so write your commit messages carefully!** Specifically, changing older commit messages requires the `git rebase` command, and editing pushed commits requires force-pushing, which we don't allow at Oppia.
+
+   * Before making the commit, do some sanity-checks:
+
+     * Ensure that your code follows [[the style rules|Coding-style-guide]].
+     * Start up a local instance of Oppia and do some manual testing in order to check that you haven't broken anything!
+     * Run `git status` to check that your changes are what you expect.  To inspect the changes made to any particular file, use `git diff your-file-name`.
+
+   * To actually make the commit, run:
+
+     ```console
+     git commit -a -m "{{YOUR COMMIT MESSAGE HERE}}"
+     ```
+
+     For example. `git commit -a -m "address nit changes"`. This will commit all the changes you've made since the last commit (shown by `git status`).
+
+     If you want to commit only some of those changes, you can use `git add` to select the changes to commit and then make a commit without the `-a` flag:
+
```

Comment:
Isn't this where the contributor is adding a new file? Just above we give the `git commit -a -m ...` command, which adds all files and makes the commit.

Also, these instructions don't say to push from a dirty branch. They just help contributors break their changes into multiple commits

Reply:
`git commit -a -m` doesn't adds untracked files, right?

Reply:
@DubeySandeep That's true. In that case, maybe we should not mention `git commit -a -m` at all since developers need to know about `git add` anyway. Further, `git commit -a -m` encourages developers to lump all their changes into one commit instead of breaking their changes up into logically chunked commits.

Here's the workflow I'd recommend:

1. Make your changes and confirm that they are correct
2. Think about how to break your changes into logical chunks
3. For each chunk, use `git add` to stage the changes in that chunk and commit them with a descriptive message

I think this will encourage better chunking of changes into commits (I don't remember where it came up, but I thought we were trying to get developers to create better commits)

Reply:
> Make your changes and confirm that they are correct
Think about how to break your changes into logical chunks
For each chunk, use git add to stage the changes in that chunk and commit them with a descriptive message

From a new contributor perspective, I would suggest having an explicit command to make a commit. (Most of the starter issues will have one commit.) Have a link or dropdown to explain best practices for commit would be better. (Note: We don't care much about commits as we do squash merge and we prefer having small PRs instead of big PRs with multiple small commits.)

Reply:
Decision from meeting: Ask contributors to run `git status`, then `git diff`, then `git add .`, and finally `git commit -m`

Reply:
done